solr/core/src/java/org/apache/solr/cli/StreamTool.java

```diff
@@ -0,0 +1,507 @@
+import java.io.StringReader;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.apache.solr.client.solrj.io.Lang;
+import org.apache.solr.client.solrj.io.SolrClientCache;
+import org.apache.solr.client.solrj.io.Tuple;
+import org.apache.solr.client.solrj.io.comp.StreamComparator;
+import org.apache.solr.client.solrj.io.stream.PushBackStream;
+import org.apache.solr.client.solrj.io.stream.SolrStream;
+import org.apache.solr.client.solrj.io.stream.StreamContext;
+import org.apache.solr.client.solrj.io.stream.TupleStream;
+import org.apache.solr.client.solrj.io.stream.expr.Explanation;
+import org.apache.solr.client.solrj.io.stream.expr.Expressible;
+import org.apache.solr.client.solrj.io.stream.expr.StreamExpression;
+import org.apache.solr.client.solrj.io.stream.expr.StreamExpressionParser;
+import org.apache.solr.client.solrj.io.stream.expr.StreamFactory;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.apache.solr.handler.CatStream;
+
+/** Supports stream command in the bin/solr script. */
+public class StreamTool extends ToolBase {
+
+  public StreamTool() {
+    this(CLIO.getOutStream());
+  }
+
+  public StreamTool(PrintStream stdout) {
+    super(stdout);
+  }
+
+  private final SolrClientCache solrClientCache = new SolrClientCache();
+
+  @Override
+  public String getName() {
+    return "stream";
+  }
+
+  private static final Option EXECUTION_OPTION =
+          Option.builder("e")
+                  .longOpt("execution")
+                  .hasArg()
+                  .argName("CONTEXT")
+                  .required(false)
+                  .desc(
+                          "Execution context is either 'local' (i.e CLI process) or 'solr'. Default is 'solr'")
+                  .build();
+
+  private static final Option COLLECTION_OPTION = Option.builder("c")
+            .longOpt("name")
+            .argName("NAME")
+            .hasArg()
+            .desc(
+                "Name of the collection to execute on if workers are 'solr'.  Required for 'solr' worker.")
+            .build();
+
+  private static final Option FIELDS_OPTION = Option.builder("f")
+            .longOpt("fields")
+            .argName("FIELDS")
+            .hasArg()
+            .required(false)
+            .desc(
+                "The fields in the tuples to output. (defaults to fields in the first tuple of result set).")
+            .build();
+
+  private static final Option HEADER_OPTION =
+          Option.builder()
+                  .longOpt("header")
+                  .required(false)
```

Comment:
Any `.required(false)` is obsolete and could be removed.

Reply:
Thanks!